Initial module version (#1)
humanmade/platform-workflow: docs/notifications.md

# Notifications

> Key to workflows is receiving timely notifications in the places where you work. This module provides a developer framework for creating notifications, either as hardcoded business logic or by making events, messages, recipients and destinations available to the Workflow builder UI.

Reviewer: ```suggestion
The key to workflows is receiving timely notifications in the places where you work. This module provides a developer framework for creating notifications, either as hardcoded business logic or by making events, messages, recipients and destinations available to the Workflow builder UI.
```

> Key to workflows is receiving timely notifications in the places where you work. This module provides a developer framework for creating notifications, either as hardcoded business logic or by making events, messages, recipients and destinations available to the Workflow builder UI.

Reviewer: ```suggestion
Key to Workflows is receiving timely notifications in the places where you work. This module provides a developer framework for creating notifications, either as hardcoded business logic or by making events, messages, recipients and destinations available to the Workflow builder UI.
```

This component is powered by the [Workflows plugin](https://github.com/humanmade/Workflows). You can find more [detailed documentation on the plugin's wiki](https://github.com/humanmade/Workflows/wiki).

> ## Basic hardcoded example

Reviewer: So can could this notification also be created via the UI? The later docs allude to be able to use the UI to construct this kind of thing. It might be good to make that explicit here?

The simplest form of a notification looks like the following:

```php
HM\Workflows\Workflow::register( 'notify_editor_on_new_post' )

>   // The action hook to trigger on.

Reviewer: What's the "action hook" is this the transition post status name? If so, maybe we could list out the options for that here.

Author: That's a lot of options! The hook is `${old_status}_to_${new_status}` so I could just put that. But it can be literally any WP action called by `do_action()`

  ->when( 'draft_to_pending' )
  // The message to be sent. Can also be a callback that recieves data from the action hook.
  ->what( 'A new post is ready for review' )
  // The user role to notify. Could also be an array of user IDs.
  ->who( 'editor' )
  // Use the 'email' destination. Can be used multiple times for multiple destinations.

>   ->where( 'email' );

Reviewer: This API is bonkers as hell! :D

Author: Yeah! Didn't think you'd be a fan 😄 Now I've had time to reflect and build on it I would like to go back to the drawing board really.

There's definite merit in the `Event` and `Destination` object implementation at least. `Destination` could really be any sort of follow on action though, not necessarily a notification.

Reviewer: Yeah, I love the functionality :D

  // action hook to allow selecting the post's author as a recipient
  // of the notification.
  ->add_recipient_handler( 'author', function ( $post_id ) {
	  return get_user_by( 'id', get_post( $post_id )->post_author );
  }, __( 'Post author' ) )
  // Add a link to the messages that are sent out. The callback can
  // return a URL to redirect the user to but can also perform any
  // actions the user clicking the link has permission to do.
  ->add_message_action(
    'view',
    __( 'View post' ),
    function ( $post_id ) {
      return get_the_permalink( $post_id );
    },
    function ( $post_id ) {
      return [ 'post_id' => $post_id ];
    },
    [ 'post_id' => 'intval' ]
  )
  // Finally add a UI, this causes the event to show in the dropdown
  // of events in the admin.
  ->add_ui( __( 'When a post is published' ) );
```

Message actions are very powerful and operate via a webhooks API under the hood. [You can learn more about the possible uses of message actions here](https://github.com/humanmade/Workflows/wiki/Event#add_message_action-string-id-string-text-stringcallable-callback_or_url-arraynull-args--null-array-schema---array-data-----event).

> Some other example uses of message actions may be for jumping to editing a post, providing a preview link, publishing a post and anything else you can think of. The links will work from any location whether in the CMS admin, an email or a slack message.

Reviewer: ```suggestion
Some other example uses of message actions may be for "jumping to editing a post", "providing a preview link", "publishing a post" and anything else you can think of. The links will work from any location whether in the CMS admin, an email or a slack message.
```

## Adding a custom destination to the UI

> Destinations handle dispatching the notifications. Email, dashboard and slack notifications are supported by default.

Reviewer: ```suggestion
Destinations handle dispatching the notifications. Email, dashboard and [Slack](https://slack.com/) notifications are supported by default.
```

The following example shows the registration of a custom destination handler that posts data to an external API using data entered via the UI.

You can create any number of integrations with external services through this method.

> **Note** the destination handler is run as a background task to prevent slowing down the site.

Reviewer: ```suggestion
**Note** the destination handler is run as a background task to prevent slowing down the script execution.
```
